0.12 crash when evaluating null inside ternary operator

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
0.12.0
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
resource "google_compute_instance_group" "foobar" {
  name = "mygroup"

  instances = [
    for server in google_compute_instance.foobar_server.*.self_link :
    substr(server,74,87) == "europe-west1-b" ? server : null
  ]
}
```

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

```Error: rpc error: code = Unavailable desc = transport is closing


panic: value is null
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4:
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: goroutine 471 [running]:
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: github.com/terraform-providers/terraform-provider-google/vendor/github.com/zclconf/go-cty/cty.Value.AsString(...)
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/github.com/zclconf/go-cty/cty/value_ops.go:1029
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/plugin/convert.PathToAttributePath(0xc003fabf40, 0x2, 0x2, 0x1)
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/plugin/convert/diagnostics.go:114 +0xa04
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/helper/plugin.validateConfigNulls(0x2fa7940, 0xc003fd2830, 0x2b938e0, 0xc003fab6e0, 0xc003fd3610, 0x1, 0x1, 0x1, 0xc003fab6e0, 0x0)
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/helper/plugin/grpc_provider.go:1312 +0x417
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/helper/plugin.validateConfigNulls(0x2fa7900, 0xc00000fc28, 0x2803fc0, 0xc003fd9200, 0x0, 0x0, 0x0, 0x0, 0xc003fd9200, 0x0)
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/helper/plugin/grpc_provider.go:1332 +0x6eb
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/helper/plugin.(*GRPCProviderServer).PlanResourceChange(0xc000162640, 0x2fa6880, 0xc003fd83f0, 0xc00070f7a0, 0xc000162640, 0xc003fd83f0, 0xc0007b8bd0)
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/helper/plugin/grpc_provider.go:595 +0x5fa
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/internal/tfplugin5._Provider_PlanResourceChange_Handler(0x2b88f60, 0xc000162640, 0x2fa6880, 0xc003fd83f0, 0xc003ef7590, 0x0, 0x2fa6880, 0xc003fd83f0, 0xc00032ec00, 0x400)
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/github.com/hashicorp/terraform/internal/tfplugin5/tfplugin5.pb.go:3001 +0x23e
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc000562780, 0x2fb5580, 0xc000556780, 0xc003fc4800, 0xc000496b40, 0x41d7328, 0x0, 0x0, 0x0)
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/grpc/server.go:971 +0x470
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/grpc.(*Server).handleStream(0xc000562780, 0x2fb5580, 0xc000556780, 0xc003fc4800, 0x0)
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/grpc/server.go:1250 +0xd25
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0000c00b0, 0xc000562780, 0x2fb5580, 0xc000556780, 0xc003fc4800)
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/grpc/server.go:690 +0x9f
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: created by github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
2019-05-28T13:01:34.343+0200 [DEBUG] plugin.terraform-provider-google_v2.7.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/vendor/google.golang.org/grpc/server.go:688 +0xa1
2019/05/28 13:01:34 [ERROR] module.compute: eval: *terraform.EvalDiff, err: rpc error: code = Unavailable desc = transport is closing
2019/05/28 13:01:34 [ERROR] module.compute: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/05/28 13:01:34 [TRACE] [walkPlan] Exiting eval tree: module.compute.google_compute_instance_group.foobar
2019/05/28 13:01:34 [TRACE] vertex "module.compute.google_compute_instance_group.foobar": visit complete
2019/05/28 13:01:34 [TRACE] vertex "module.compute.google_compute_instance_group.foobar": dynamic subgraph encountered errors
2019/05/28 13:01:34 [TRACE] vertex "module.compute.google_compute_instance_group.foobar": visit complete
2019/05/28 13:01:34 [TRACE] dag/walk: upstream of "module.compute.google_compute_backend_service.foobar_backend_service" errored, so skipping
2019/05/28 13:01:34 [TRACE] dag/walk: upstream of "module.compute.google_compute_url_map.foobar_url_map" errored, so skipping
2019/05/28 13:01:34 [TRACE] dag/walk: upstream of "module.compute.google_compute_target_https_proxy.foobar_https_proxy" errored, so skipping
2019/05/28 13:01:34 [TRACE] dag/walk: upstream of "module.compute.google_compute_global_forwarding_rule.xb_foobar_forwarding_rule" errored, so skipping
2019-05-28T13:01:34.353+0200 [DEBUG] plugin: plugin process exited: path=**redacted**.terraform/plugins/darwin_amd64/terraform-provider-google_v2.7.0_x4 pid=89114 error="exit status 2"
2019/05/28 13:01:34 [TRACE] dag/walk: upstream of "provider.google (close)" errored, so skipping
2019/05/28 13:01:34 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2019/05/28 13:01:34 [TRACE] dag/walk: upstream of "root" errored, so skipping
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=89105
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=89108
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.532+0200 [ERROR] plugin.terraform: reading plugin stderr: error="read |0: file already closed"
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=88973
2019-05-28T13:01:34.532+0200 [ERROR] plugin.terraform: reading plugin stderr: error="read |0: file already closed"
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=88972
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=88968
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=88970
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=89107
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.532+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.534+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=88969
2019-05-28T13:01:34.534+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.534+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=89106
2019-05-28T13:01:34.534+0200 [DEBUG] plugin: plugin exited
2019-05-28T13:01:34.534+0200 [DEBUG] plugin: plugin process exited: path=/usr/local/bin/terraform pid=89110
2019-05-28T13:01:34.534+0200 [DEBUG] plugin: plugin exited
```

### Expected Behavior
<!--
What should have happened?
-->
The idea is to loop over the `google_compute_instance.foobar_server.*.self_link` list and only return values that match a specific criteria, `null` otherwise ( `""` instead of `null` is not an option since `instances=[]` expects a valid self_link or nothing). The use of `null` here is essentially a hack to skip certain loop steps.
Please note that if `google_compute_instance.foobar_server.*.self_link` is empty ie. no `google_compute_instance.foobar_server` was created, the plan will suceed and `instances = []` will be empty as expected.
### Actual Behavior
<!--
What actually happened?
-->
Crash

--- a/helper/plugin/grpc_provider.go
+++ b/helper/plugin/grpc_provider.go
@@ -1374,11 +1374,19 @@ func validateConfigNulls(v cty.Value, path cty.Path) []*proto.Diagnostic {
 		for it.Next() {
 			kv, ev := it.Element()
 			if ev.IsNull() {
+				// if this is a set, the kv is also going to be null which
+				// isn't a valid path element, so we can't append it to the
+				// diagnostic.
+				p := path
+				if !kv.IsNull() {
+					p = append(p, cty.IndexStep{Key: kv})
+				}
+
 				diags = append(diags, &proto.Diagnostic{
 					Severity:  proto.Diagnostic_ERROR,
 					Summary:   "Null value found in list",
 					Detail:    "Null values are not allowed for this attribute value.",
-					Attribute: convert.PathToAttributePath(append(path, cty.IndexStep{Key: kv})),
+					Attribute: convert.PathToAttributePath(p),
 				})
 				continue
 			}

--- a/helper/plugin/grpc_provider_test.go
+++ b/helper/plugin/grpc_provider_test.go
@@ -1352,6 +1352,17 @@ func TestValidateNulls(t *testing.T) {
 			}),
 			Err: true,
 		},
+		{
+			Cfg: cty.ObjectVal(map[string]cty.Value{
+				"object": cty.ObjectVal(map[string]cty.Value{
+					"list": cty.SetVal([]cty.Value{
+						cty.StringVal("string"),
+						cty.NullVal(cty.String),
+					}),
+				}),
+			}),
+			Err: true,
+		},
 	} {
 		t.Run(strconv.Itoa(i), func(t *testing.T) {
 			d := validateConfigNulls(tc.Cfg, nil)

